src/pymatgen/util/string.py

```diff
@@ -97,57 +107,65 @@ def to_unicode_string(self):
         return str_
 
 
-def str_delimited(results, header=None, delimiter="\t"):
+def str_delimited(
+    results: Sequence[Sequence],
+    header: Sequence[str] | None = None,
+    delimiter: str = "\t",
+) -> str:
     r"""Given a tuple of tuples, generate a delimited string form.
-    >>> results = [["a", "b", "c"], ["d", "e", "f"], [1, 2, 3]]
+    >>> results = (("a", "b", "c"), ("d", "e", "f"), (1, 2, 3))
     >>> print(str_delimited(results, delimiter=","))
-    a,b,c
-    d,e,f
-    1,2,3.
+        a,b,c
+        d,e,f
+        1,2,3.
 
     Args:
-        results: 2d sequence of arbitrary types.
-        header: optional header
-        delimiter: Defaults to "\t" for tab-delimited output.
+        results (Sequence[Sequence]): 2d sequence of arbitrary types.
+        header (Sequence[str]): optional headers.
+        delimiter (str): Defaults to "\t" for tab-delimited output.
 
     Returns:
-        Aligned string output in a table-like format.
+        str: Aligned string output in a table-like format.
     """
     out = ""
     if header is not None:
         out += f"{delimiter.join(header)}\n"
     return out + "\n".join(delimiter.join([str(m) for m in result]) for result in results)
 
 
-def formula_double_format(afloat, ignore_ones=True, tol: float = 1e-8):
-    """This function is used to make pretty formulas by formatting the amounts.
-    Instead of Li1.0 Fe1.0 P1.0 O4.0, you get LiFePO4.
+def formula_double_format(
+    afloat: float,
+    ignore_ones: bool = True,
+    tol: float = 1e-8,
+) -> float | str:
+    """Make pretty formulas by formatting the amounts.
+    E.g. "Li1.0 Fe1.0 P1.0 O4.0" -> "LiFePO4".
 
     Args:
-        afloat (float): a float
-        ignore_ones (bool): if true, floats of 1 are ignored.
-        tol (float): Tolerance to round to nearest int. i.e. 2.0000000001 -> 2
+        afloat (float): a float.
+        ignore_ones (bool): if true, floats of 1.0 are ignored.
+        tol (float): Absolute tolerance to round to nearest int. i.e. 2.0000000001 -> 2.
 
     Returns:
-        A string representation of the float for formulas.
+        str | float: A string representation of the float for formulas.
     """
-    if ignore_ones and afloat == 1:
+    if ignore_ones and math.isclose(afloat, 1, abs_tol=tol):
         return ""
-    if abs(afloat - int(afloat)) < tol:
+    if abs(afloat - int(afloat)) <= tol:
         return int(afloat)
```

Comment:
Here is a bug, `int()` would not round to "nearest" integer, it would just truncates the decimal part:
```python
print(int(2.9))  # 2
```